Doxygen (#78)
bmwcarit/MoCOCrW: doc/examples/example_ecies.md

# Elliptic Curve Integrated Encryption Scheme (ECIES)

# Basic ECIES

ECIES utilizes well-known symmetric encryption, key derivation functions and message authentication codes
to achieve encryption based on elliptic curves. However, ECIES is not limited to specific algorithms. That is,

> the algorithms used for symmetric encryption, key derivation and message authentication can be configured.

Reviewer: configured as in? I guess this should be that it can be configured in our implementation?

For the sake of easy usage, we have decided to use the following defaults:

>  - Symmetric Encryption: AES-CBC with PKCS7 padding, IV=0x0

Reviewer: key size?

 - Key Derivation Function: X9.63 without salt
 - Message Authentication Code: HMAC-SHA512 without salt

> The following examples show to en- and decrypt using ECIES with these default settings. Afterwards the

Reviewer: ```suggestion
The following examples show how to en- and decrypt using ECIES with these default settings. Afterwards, the
```

different options to change the default values will be shown.

## Encryption

```cpp
// Generate Key pair

> mococrw::AsymmetricKeypair secp384Key = mococrw::AsymmetricKeypair::generate(mococrw::ECCSpec(mococrw::openssl::ellipticCurveNid::SECP_384r1));

Reviewer: why do you generate a keypair here and load it from PEM below? In a standard usage scenario, you would load the public key from PEM here as well…

mococrw::AsymmetricPublicKey secp384PublicKey = secp384Key;

std::vector<uint8_t> testString = {'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd', '!'};

mococrw::ECIESCtxBuilder encBuilder;
std::unique_ptr<mococrw::ECIESEncryptionCtx> encCtx = encBuilder.buildEncryptionCtx(secp384PublicKey);
encCtx->update(testString);

// Finish encryption and extract parameters for decryption
std::vector<uint8_t> ciphertext = encCtx->finish();
std::vector<uint8_t> mac = encCtx->getMAC();
mococrw::AsymmetricPublicKey ephKey = encCtx->getEphemeralKey();
```

## Decryption

```cpp
mococrw::AsymmetricPublicKey ephKey = mococrw::AsymmetricPublicKey::readPublicKeyFromPEM(ephKeyPEM);
mococrw::AsymmetricKeypair privKey = mococrw::AsymmetricKeypair::readPrivateKeyFromPEM(secp384KeyPEM, "password");
std::vector<uint8_t> ciphertext = { ... };
std::vector<uint8_t> mac = { ... };

mococrw::ECIESCtxBuilder decBuilder;

// Set parameters for decryption

> std::unique_ptr<mococrw::ECIESDecryptionCtx> decCtx = decBuilder.buildDecryptionCtx(privKey, ephKey);

Reviewer: This ephermeral key should get more attention. My understanding here is that this key is usually not transported as PEM but as raw curve point. Our example should probably reflect this. Also in the encryption example. Maybe this is also a serialization topic that we should comment on? s.b.…

Author: agree, changed the example to not use PEM anymore. To me seems like the API is a bit inconvenient:
`mococrw::AsymmetricKeypair::getKeySpec()` returns a `std::unique_ptr<mococrw::AsymmetricKey::Spec>` but `mococrw::AsymmetricPublicKey::fromECPoint()` expects a `std::shared_ptr<mococrw::ECCSpec>`, which means we need to convert the unique_ptr into a shared_ptr and also cast it. Or am I missing the reason why we are using unique_ptr and shared_ptr for the spec types here?

decCtx->update(ciphertext);
decCtx->setMAC(mac);

// Decrypt
try {
    std::vector<uint8_t> result = decCtx->finish();
}
catch (const MoCOCrWException &e)  {
    std::cerr << "MAC verification failed" << std:endl;
    ...
}

> ```

Reviewer: There should be a comment regarding the serialization format. In particular, it should state that the library wants to make no assumptions for this and the user of the library is responsible for that.

# Change Default Parameters

## Symmetric Cipher
The `mococrw::ECIESCtxBuilder` offers methods to modify the symmetric cipher that will be used for encryption.
The following code shows how to use these methods to change the symmetric cipher to AES-CBC with a key size of 128 bit:

### Encryption

```cpp
// Generate Key pair
mococrw::AsymmetricKeypair secp384Key = mococrw::AsymmetricKeypair::generate(mococrw::ECCSpec(mococrw::openssl::ellipticCurveNid::SECP_384r1));
mococrw::AsymmetricPublicKey secp384PublicKey = secp384Key;

std::vector<uint8_t> testString = {'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd', '!'};

mococrw::ECIESCtxBuilder encBuilder;

// 1. Set new Symmetric Cipher Factory Function
auto cipherEncFunc = [](const std::vector<uint8_t> &key) -> std::unique_ptr<mococrw::AESCipher> {
    return mococrw::AESCipherBuilder(mococrw::SymmetricCipherMode::CBC, mococrw::SymmetricCipherKeySize::S_128, key)

>             .setIV(std::vector<uint8_t>(mococrw::AESCipherBuilder::getDefaultIVLength(mococrw::SymmetricCipherMode::CBC)))

Reviewer: we feed the IV length in the IV?

Author: well, this generates a std::vector with the length of the IV. Actually i've copied this from the doxygen documentation (and didn't pay to close attention). But agree that this is strange..

Reviewer: true. I must have missed the std::vector constructor in this line. Well, though it probably works then, I guess this is a clear sign that this is not easily readable code and should be improved.

};
encBuilder.setSymmetricCipherFactoryFunction(cipherEncFunc);

// 2. Set new key size
encBuilder.setSymmetricCipherKeySize(mococrw::getSymmetricCipherKeySize(mococrw::SymmetricCipherKeySize::S_128));

std::unique_ptr<mococrw::ECIESEncryptionCtx> encCtx = encBuilder.buildEncryptionCtx(secp384PublicKey);
encCtx->update(testString);

// Finish encryption and extract parameters for decryption
std::vector<uint8_t> ciphertext = encCtx->finish();
std::vector<uint8_t> mac = encCtx->getMAC();
mococrw::AsymmetricPublicKey ephKey = encCtx->getEphemeralKey();
```

### Decryption

```cpp
std::string pemKey =
mococrw::AsymmetricPublicKey ephKey = mococrw::AsymmetricPublicKey::readPublicKeyFromPEM(ephKeyPEM);
mococrw::AsymmetricKeypair privKey = mococrw::AsymmetricKeypair::readPrivateKeyFromPEM(secp384KeyPEM, "");
std::vector<uint8_t> ciphertext = { ... };
std::vector<uint8_t> mac = { ... };

mococrw::ECIESCtxBuilder decBuilder;

// 1. Set new Symmetric Cipher Factory Function
auto cipherDecFunc = [](const std::vector<uint8_t> &key) -> std::unique_ptr<mococrw::AESCipher> {
    return mococrw::AESCipherBuilder(mococrw::SymmetricCipherMode::CBC, mococrw::SymmetricCipherKeySize::S_128, key)

>             .setIV(std::vector<uint8_t>(mococrw::AESCipherBuilder::getDefaultIVLength(mococrw::SymmetricCipherMode::CBC)))

Reviewer: s.a.

            .setPadding(mococrw::SymmetricCipherPadding::PKCS)
            .buildDecryptor();
};
decBuilder.setSymmetricCipherFactoryFunction(cipherDecFunc);

// 2. Set new key size
decBuilder.setSymmetricCipherKeySize(mococrw::getSymmetricCipherKeySize(mococrw::SymmetricCipherKeySize::S_128));

// Set parameters for decryption
std::unique_ptr<mococrw::ECIESDecryptionCtx> decCtx = decBuilder.buildDecryptionCtx(privKey, ephKey);
decCtx->update(ciphertext);
decCtx->setMAC(mac);

// Decrypt
try {
    std::vector<uint8_t> result = decCtx->finish();
}
catch (const MoCOCrWException &e)  {
    std::cerr << "MAC verification failed" << std:endl;
    ...
}
```

## Key Derivation Function (KDF)

> MoCOCrW offers different key derivation functions. The following example show how to change the KDF used by ECIES. In addition, the example show

Reviewer: ```suggestion
MoCOCrW offers different key derivation functions. The following example show how to change the KDF used by ECIES. In addition, the example shows
```

# Elliptic Curve Integrated Encryption Scheme (ECIES)

# Basic ECIES

ECIES utilizes well-known symmetric encryption, key derivation functions and message authentication codes
to achieve encryption based on elliptic curves. However, ECIES is not limited to specific algorithms. That is,
the algorithms used for symmetric encryption, key derivation and message authentication can be configured

> in out implementation. For the sake of easy usage, we have decided to use the following defaults:

Reviewer 2: ```suggestion
in our implementation. For the sake of easy usage, we have decided to use the following defaults:
```

std::vector<uint8_t> point = { ... };

// 1. Read private key from PEM
mococrw::AsymmetricKeypair privKey = mococrw::AsymmetricKeypair::readPrivateKeyFromPEM(secp384KeyPEM, "password");

// 2. Get type of curve from private key
std::shared_ptr<mococrw::ECCSpec> spec;
{
    // Convert returned unique_ptr into shared_ptr
    std::shared_ptr<mococrw::AsymmetricKey::Spec> spec_tmp = privKey.getKeySpec();

    // Cast shared_ptr<mococrw::AsymmetricKey::Spec> into std::shared_ptr<mococrw::ECCSpec>
    spec = std::dynamic_pointer_cast<mococrw::ECCSpec>(spec_tmp);
}

if (!spec) {
    // Error handling
    ...
}

// 3. Re-create ephemeral key
mococrw::AsymmetricPublicKey ephKey = mococrw::AsymmetricPublicKey::fromECPoint(spec, point);

mococrw::ECIESCtxBuilder decBuilder;

// 4. Set parameters for decryption
std::unique_ptr<mococrw::ECIESDecryptionCtx> decCtx = decBuilder.buildDecryptionCtx(privKey, ephKey);
decCtx->update(ciphertext);
decCtx->setMAC(mac);

// 5. Decrypt
try {
    std::vector<uint8_t> result = decCtx->finish();
}
catch (const mococrw::MoCOCrWException &e)  {
    std::cerr << "MAC verification failed" << std::endl;
    ...
}
```

## Serialization of Artifacts

> The encrption of the plaintext creates several artiacts (ciphertext, mac and ephemeral key) that need to be transfered to the client in order to

Reviewer 2: ```suggestion
The encryption of the plaintext creates several artifacts (ciphertext, mac and ephemeral key) that need to be transferred to the client in order to
```

> decrypt the ciphertext. MoCOCrW offer no serilaization functionality because we don't want to make any assumptions of the format. That is,

Reviewer 2: ```suggestion
decrypt the ciphertext. MoCOCrW offers no serialization functionality because we don't want to make any assumptions of the format. That is,
```

It is also possible to encode the ephemeral key as PEM, but this seems rather unsusal.

# Change Default Parameters

## Symmetric Cipher
The `mococrw::ECIESCtxBuilder` offers methods to modify the symmetric cipher that will be used for encryption.
The following code shows how to use these methods to change the symmetric cipher to AES-CBC with a key size of 128 bit:

### Encryption

```cpp
// 1. Read public key from PEM
mococrw::AsymmetricPublicKey pubKey = mococrw::AsymmetricPublicKey::readPublicKeyFromPEM(secp384PubKeyPEM);

std::vector<uint8_t> testString = {'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd', '!'};

mococrw::ECIESCtxBuilder encBuilder;

// 2. Create IV vector (default value for the IV is 0x0 in ECIES)
// Default size of the IV can be obtianed with mococrw::AESCipherBuilder::getDefaultIVLength(mococrw::SymmetricCipherMode::CBC)
std::vector<uint8_t> IV = ...

// 3. Set new Symmetric Cipher Factory Function
auto cipherEncFunc = [](const std::vector<uint8_t> &key) -> std::unique_ptr<mococrw::AESCipher> {
    return mococrw::AESCipherBuilder(mococrw::SymmetricCipherMode::CBC, mococrw::SymmetricCipherKeySize::S_128, key)

>             .setIV(std::vector<uint8_t>(IV))

Reviewer 2: This will require capturing IV in the lambda, no?

// Ciphertext, mac, point and IV need to be available
std::vector<uint8_t> ciphertext = { ... };
std::vector<uint8_t> mac = { ... };
std::vector<uint8_t> point = { ... };
std::vector<uint8_t> IV = { ... };

// 1. Read private key from PEM
mococrw::AsymmetricKeypair privKey = mococrw::AsymmetricKeypair::readPrivateKeyFromPEM(secp384KeyPEM, "");

// 2. Get type of curve from private key
std::shared_ptr<mococrw::ECCSpec> spec;
{
    // Convert returned unique_ptr into shared_ptr
    std::shared_ptr<mococrw::AsymmetricKey::Spec> spec_tmp = privKey.getKeySpec();

    // Cast shared_ptr<mococrw::AsymmetricKey::Spec> into std::shared_ptr<mococrw::ECCSpec>
    spec = std::dynamic_pointer_cast<mococrw::ECCSpec>(spec_tmp);
}

if (!spec) {
    // Error handling
    ...
}

// 3. Re-create ephemeral key
mococrw::AsymmetricPublicKey ephKey = mococrw::AsymmetricPublicKey::fromECPoint(spec, point);

mococrw::ECIESCtxBuilder decBuilder;

// 4. Set new Symmetric Cipher Factory Function
auto cipherDecFunc = [](const std::vector<uint8_t> &key) -> std::unique_ptr<mococrw::AESCipher> {
    return mococrw::AESCipherBuilder(mococrw::SymmetricCipherMode::CBC, mococrw::SymmetricCipherKeySize::S_128, key)

>             .setIV(IV)

Reviewer 2: Same here, capture IV

            .setPadding(mococrw::SymmetricCipherPadding::PKCS)
            .buildDecryptor();
};
decBuilder.setSymmetricCipherFactoryFunction(cipherDecFunc);

// 5. Set new key size
decBuilder.setSymmetricCipherKeySize(mococrw::getSymmetricCipherKeySize(mococrw::SymmetricCipherKeySize::S_128));

// 6. Set parameters for decryption
std::unique_ptr<mococrw::ECIESDecryptionCtx> decCtx = decBuilder.buildDecryptionCtx(privKey, ephKey);
decCtx->update(ciphertext);
decCtx->setMAC(mac);

// 7. Decrypt
try {
    std::vector<uint8_t> result = decCtx->finish();
}
catch (const MoCOCrWException &e)  {
    std::cerr << "MAC verification failed" << std::endl;
    ...
}
```

## Key Derivation Function (KDF)

> MoCOCrW offers different key derivation functions. The following example show how to change the KDF used by ECIES. In addition, the example shows

Reviewer 2: ```suggestion
MoCOCrW offers different key derivation functions. The following example shows how to change the KDF used by ECIES. In addition, the example shows
```

This introduces a repetition of "examples shows how to", so maybe we can change one of them to "the example demonstrates setting …"

std::shared_ptr<mococrw::ECCSpec> spec;
{
    // Convert returned unique_ptr into shared_ptr
    std::shared_ptr<mococrw::AsymmetricKey::Spec> spec_tmp = privKey.getKeySpec();

    // Cast shared_ptr<mococrw::AsymmetricKey::Spec> into std::shared_ptr<mococrw::ECCSpec>
    spec = std::dynamic_pointer_cast<mococrw::ECCSpec>(spec_tmp);
}

if (!spec) {
    // Error handling
    ...
}

// 3. Re-create ephemeral key
mococrw::AsymmetricPublicKey ephKey = mococrw::AsymmetricPublicKey::fromECPoint(spec, point);

mococrw::ECIESCtxBuilder decBuilder;

// 4. Set KDF to PBKDF2
decBuilder.setKDF(std::make_shared<mococrw::PBKDF2>(mococrw::openssl::DigestTypes::SHA256, 100));

// (optionally) Set salt for KDF
std::vector<uint8_t> encryptionSalt= { ... };
encBuilder.setKDFSalt(encryptionSalt);

// 5. Set parameters for decryption
std::unique_ptr<mococrw::ECIESDecryptionCtx> decCtx = decBuilder.buildDecryptionCtx(privKey, ephKey);
decCtx->update(ciphertext);
decCtx->setMAC(mac);

// 6. Decrypt
try {
    std::vector<uint8_t> result = decCtx->finish();
}
catch (const MoCOCrWException &e)  {
    std::cerr << "MAC verification failed" << std::endl;
    ...
}
```

## Message Authentication Code (MAC)

> The following example shows how to configure (and specify if there every shall be more one supported MAC) the MAC to be used by the ECIES scheme

Reviewer 2: ```suggestion
The following example shows how to configure (and specify if there ever shall be more one supported MAC) the MAC to be used by the ECIES scheme
```

> The following example shows how to configure (and specify if there ever shall be more one supported MAC) the MAC to be used by the ECIES scheme

Reviewer 2: ```suggestion
The following example shows how to configure (and specify if there ever shall be more than one supported MAC) the MAC to be used by the ECIES scheme
```
